From d8929ea790dca6868fdb436baa779e653c70e36e Mon Sep 17 00:00:00 2001
From: Yuhao Xiang <xiangyuhao2@huawei.com>
Date: Thu, 18 Sep 2025 22:31:41 +0800
Subject: [PATCH 809/960] downstream: ub: ubfi: Support builtin parameter
 modification

ub:ubus: Support for ub bus driver framework
ub:ubfi: Support ub firmware interface basic functions
ub:ubfi: ubfi driver parse ubc information from ubrt
ub:ubfi: ubfi driver create ubc device
ub: ubfi: Parsing ummu node in the ubrt table
ub:ubfi: register ubc usi domain
ub:ubus: Fix X86 build error
ub:hisi-ubus ub:hisi-ubus: Move the decoder's page table operations to hisi-ubus
iommu/ummu: Remove redundant CONFIG_UB_UBRT_PLAT_DEV and code
iommu/ummu: Delete the unused interface of the UMMU

Signed-off-by: Yuhao Xiang <xiangyuhao2@huawei.com>
Signed-off-by: huwentao <huwentao19@h-partners.com>
---
 drivers/irqchip/irq-gic-v3-its-platform-msi.c |   1 +
 drivers/ub/Makefile                           |   3 +
 drivers/ub/ubfi/Makefile                      |   3 +
 drivers/ub/ubfi/irq.c                         |  79 +++++++
 drivers/ub/ubfi/ubrt-fwnode.c                 | 131 +++++++++++
 include/ub/ubfi/ubfi.h                        | 217 ++++++++++++++++++
 6 files changed, 434 insertions(+)
 create mode 100644 drivers/ub/Makefile
 create mode 100644 drivers/ub/ubfi/Makefile
 create mode 100644 drivers/ub/ubfi/irq.c
 create mode 100644 drivers/ub/ubfi/ubrt-fwnode.c
 create mode 100644 include/ub/ubfi/ubfi.h

diff --git a/drivers/irqchip/irq-gic-v3-its-platform-msi.c b/drivers/irqchip/irq-gic-v3-its-platform-msi.c
index 7334598ec133..35d0dd74aeb1 100644
--- a/drivers/irqchip/irq-gic-v3-its-platform-msi.c
+++ b/drivers/irqchip/irq-gic-v3-its-platform-msi.c
@@ -9,6 +9,7 @@
 #include <linux/msi.h>
 #include <linux/of.h>
 #include <linux/of_irq.h>
+#include <ub/ubfi/ubfi.h>
 
 #ifdef CONFIG_VIRT_PLAT_DEV
 static struct irq_domain *vp_irq_domain;
diff --git a/drivers/ub/Makefile b/drivers/ub/Makefile
new file mode 100644
index 000000000000..7e219469dd02
--- /dev/null
+++ b/drivers/ub/Makefile
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: GPL-2.0+
+
+obj-y += ubfi/
diff --git a/drivers/ub/ubfi/Makefile b/drivers/ub/ubfi/Makefile
new file mode 100644
index 000000000000..effdc869e58e
--- /dev/null
+++ b/drivers/ub/ubfi/Makefile
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: GPL-2.0+
+
+obj-y			+= irq.o ubrt-fwnode.o
diff --git a/drivers/ub/ubfi/irq.c b/drivers/ub/ubfi/irq.c
new file mode 100644
index 000000000000..3f449ff5aa0b
--- /dev/null
+++ b/drivers/ub/ubfi/irq.c
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (c) HiSilicon Technologies Co., Ltd. 2025. All rights reserved.
+ */
+
+#include <linux/acpi_iort.h>
+#include <linux/irqdomain.h>
+#include <linux/msi.h>
+#include <linux/platform_device.h>
+#include <ub/ubfi/ubfi.h>
+
+int ub_update_msi_domain(struct device *dev,
+			 enum irq_domain_bus_token bus_token)
+{
+#ifdef CONFIG_GENERIC_MSI_IRQ
+	struct fwnode_handle *fwnode;
+	struct irq_domain *domain;
+
+	domain = dev->msi.domain;
+	if (!domain) {
+		dev_err(dev, "find base irq domain failed!\n");
+		return -ENODEV;
+	}
+
+	fwnode = domain->fwnode;
+	if (!fwnode) {
+		dev_err(dev, "find fwnode failed!\n");
+		return -ENODEV;
+	}
+
+	domain = irq_find_matching_fwnode(fwnode, bus_token);
+	if (!domain) {
+		dev_err(dev, "find irq domain failed!\n");
+		return -ENODEV;
+	}
+
+	/* Update msi domain with new bus_token */
+	dev_set_msi_domain(dev, domain);
+#endif
+	return 0;
+}
+EXPORT_SYMBOL_GPL(ub_update_msi_domain);
+
+int ubrt_register_gsi(u32 hwirq, int trigger, int polarity, const char *name,
+		      struct resource *res)
+{
+#ifdef CONFIG_ACPI
+	int irq;
+
+	if (!res) {
+		pr_err("ub register gsi, res is null\n");
+		return -EINVAL;
+	}
+
+	irq = acpi_register_gsi(NULL, hwirq, trigger, polarity);
+	if (irq <= 0) {
+		pr_err("could not register ub gsi hwirq %u\n", hwirq);
+		return -EINVAL;
+	}
+
+	res->name = name;
+	res->start = irq;
+	res->end = irq;
+	res->flags = IORESOURCE_IRQ;
+	return 0;
+#else
+	return -EINVAL;
+#endif
+}
+EXPORT_SYMBOL_GPL(ubrt_register_gsi);
+
+void ubrt_unregister_gsi(u32 hwirq)
+{
+#ifdef CONFIG_ACPI
+	acpi_unregister_gsi(hwirq);
+#endif
+}
+EXPORT_SYMBOL_GPL(ubrt_unregister_gsi);
+
diff --git a/drivers/ub/ubfi/ubrt-fwnode.c b/drivers/ub/ubfi/ubrt-fwnode.c
new file mode 100644
index 000000000000..1e5f3f6f1fcb
--- /dev/null
+++ b/drivers/ub/ubfi/ubrt-fwnode.c
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (c) HiSilicon Technologies Co., Ltd. 2025. All rights reserved.
+ * Descriptor: Maintain ubrt information acquisition path
+ */
+#include <linux/fwnode.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+#include <linux/spinlock.h>
+#include <ub/ubfi/ubfi.h>
+
+static LIST_HEAD(ubrt_fwnode_list);
+static DEFINE_SPINLOCK(ubrt_fwnode_lock);
+
+int ubrt_fwnode_set(u32 index, enum ubrt_node_type type, struct fwnode_handle *fwnode)
+{
+	struct ubrt_fwnode *curr;
+
+	if (!fwnode)
+		return -ENODEV;
+
+	guard(spinlock)(&ubrt_fwnode_lock);
+	list_for_each_entry(curr, &ubrt_fwnode_list, list) {
+		if (curr->type == type && curr->index == index) {
+			curr->fwnode = fwnode;
+			return 0;
+		}
+	}
+
+	return -ENXIO;
+}
+EXPORT_SYMBOL_GPL(ubrt_fwnode_set);
+
+struct ubrt_fwnode *ubrt_fwnode_get_by_idx(u32 index, enum ubrt_node_type type)
+{
+	struct ubrt_fwnode *curr;
+
+	guard(spinlock)(&ubrt_fwnode_lock);
+	list_for_each_entry(curr, &ubrt_fwnode_list, list) {
+		if (curr->type == type && curr->index == index)
+			return curr;
+	}
+	return NULL;
+}
+EXPORT_SYMBOL_GPL(ubrt_fwnode_get_by_idx);
+
+struct ubrt_fwnode *ubrt_fwnode_get(struct fwnode_handle *fwnode)
+{
+	struct ubrt_fwnode *curr;
+
+	guard(spinlock)(&ubrt_fwnode_lock);
+	list_for_each_entry(curr, &ubrt_fwnode_list, list) {
+		if (curr->fwnode == fwnode)
+			return curr;
+
+	}
+	return NULL;
+}
+EXPORT_SYMBOL_GPL(ubrt_fwnode_get);
+
+u32 ubrt_fwnode_get_count(enum ubrt_node_type type)
+{
+	struct ubrt_fwnode *curr;
+	u32 count = 0;
+
+	guard(spinlock)(&ubrt_fwnode_lock);
+	list_for_each_entry(curr, &ubrt_fwnode_list, list) {
+		if (curr->type == type)
+			count++;
+	}
+	return count;
+}
+EXPORT_SYMBOL_GPL(ubrt_fwnode_get_count);
+
+int ubrt_fwnode_add(void *node, u32 index, int size, enum ubrt_node_type type)
+{
+	struct ubrt_fwnode *np;
+	void *tmp;
+
+	if (!node)
+		return -EINVAL;
+
+	np = kzalloc(sizeof(*np), GFP_KERNEL);
+	if (!np)
+		return -ENOMEM;
+
+	tmp = kmemdup(node, size, GFP_KERNEL);
+	if (!tmp) {
+		kfree(np);
+		return -ENOMEM;
+	}
+
+	INIT_LIST_HEAD(&np->list);
+	np->ubrt_node = tmp;
+	np->index = index;
+	np->type = type;
+	guard(spinlock)(&ubrt_fwnode_lock);
+	list_add_tail(&np->list, &ubrt_fwnode_list);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(ubrt_fwnode_add);
+
+void ubrt_fwnode_del(u32 index, enum ubrt_node_type type)
+{
+	struct ubrt_fwnode *curr, *tmp;
+
+	guard(spinlock)(&ubrt_fwnode_lock);
+	list_for_each_entry_safe(curr, tmp, &ubrt_fwnode_list, list) {
+		if (curr->type == type && curr->index == index) {
+			kfree(curr->ubrt_node);
+			list_del(&curr->list);
+			kfree(curr);
+			break;
+		}
+	}
+}
+EXPORT_SYMBOL_GPL(ubrt_fwnode_del);
+
+void ubrt_fwnode_del_all(void)
+{
+	struct ubrt_fwnode *cur_fwnode, *tmp_fwnode;
+
+	guard(spinlock)(&ubrt_fwnode_lock);
+	list_for_each_entry_safe(cur_fwnode, tmp_fwnode, &ubrt_fwnode_list, list) {
+		kfree(cur_fwnode->ubrt_node);
+		list_del(&cur_fwnode->list);
+		kfree(cur_fwnode);
+	}
+}
+EXPORT_SYMBOL_GPL(ubrt_fwnode_del_all);
diff --git a/include/ub/ubfi/ubfi.h b/include/ub/ubfi/ubfi.h
new file mode 100644
index 000000000000..a5d1019d80fe
--- /dev/null
+++ b/include/ub/ubfi/ubfi.h
@@ -0,0 +1,217 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Copyright (c) HiSilicon Technologies Co., Ltd. 2025. All rights reserved.
+ */
+
+#ifndef _UB_UBFI_UBFI_H_
+#define _UB_UBFI_UBFI_H_
+
+#include <linux/device.h>
+#include <linux/ioport.h>
+#include <linux/irqdomain_defs.h>
+#include <linux/list.h>
+#include <linux/types.h>
+
+#define UMMU_RSV_LEN 26
+#define UMMU_VEND_LEN 10
+
+enum ubrt_node_type {
+	UBRT_UBC,
+	UBRT_UMMU,
+	UBRT_UMMU_PMU,
+};
+
+/*
+ * struct ubrt_fwnode - Structure to store nodes of the UBRT table reported by
+ *                      the BIOS. The UBRT nodes describe hardware information
+ *                      for UB system devices, such as UBC devices, UMMU
+ *                      devices, etc.
+ *
+ * @type: Type of the UBRT node.
+ * @index: Index of the UBRT node within its type.
+ * @fwnode: fwnode_handle of the device corresponding to the UBRT node.
+ * @ubrt_node: Pointer to the UBRT node, e.g., an ummu_node.
+ * @list: List node of ubrt_fwnode_list.
+ */
+struct ubrt_fwnode {
+	enum ubrt_node_type type;
+	u32 index;
+	struct fwnode_handle *fwnode;
+	void *ubrt_node;
+	struct list_head list;
+};
+
+/*
+ * struct ummu_node - Structure to store hardware information nodes reported
+ *                    by the BIOS to the UMMU device.
+ *
+ * @base_addr: Base address of the UMMU register space.
+ * @addr_size: Size of the UMMU register space.
+ * @intr_id: Interrupt ID for the UMMU device, used for device interrupt
+ *           request in the OS.
+ * @pxm: Proximity Domain (PXM) identifier of the UMMU device. If the value is
+ *       0xFFFF, it indicates an invalid PXM.
+ * @its_index: Index of the Interrupt Translation Service (ITS) associated with
+ *             the UMMU device.
+ * @pmu_addr: Base address of the PMU register space associated with the UMMU
+ *            device.
+ * @pmu_size: Size of the PMU register space associated with the UMMU device.
+ * @pmu_intr_id: Interrupt ID for the PMU associated with the UMMU device.
+ * @min_tid: Minimum Token ID that can be allocated.
+ * @max_tid: Maximum Token ID that can be allocated.
+ * @reserved: Reserved bytes.
+ * @vendor_id: Vendor ID, reused from the Vendor ID definition in GUID, see
+ *             "UB Base Specification 2.0".
+ * @vendor_info: Vendor-specific information.
+ */
+struct ummu_node {
+	u64 base_addr;
+	u64 addr_size;
+	u32 intr_id;
+	u16 pxm;
+	u16 its_index;
+	u64 pmu_addr;
+	u64 pmu_size;
+	u32 pmu_intr_id;
+	u32 min_tid;
+	u32 max_tid;
+	u8 reserved[UMMU_RSV_LEN];
+	u16 vendor_id;
+	u64 vendor_info[UMMU_VEND_LEN];
+};
+
+/**
+ * ubrt_register_gsi() - Registering UBC's interrupt into the kernel via ACPI
+ * @hwirq: GSI IRQ number reported by BIOS
+ * @trigger: trigger type of the GSI number to be mapped
+ * @polarity: polarity of the GSI to be mapped
+ * @name: GSI IRQ name
+ * @res: Record the soft interrupt number obtained after registration into res
+ *
+ * Return: 0 if success or other if failed
+ */
+int ubrt_register_gsi(u32 hwirq, int trigger, int polarity, const char *name,
+		      struct resource *res);
+
+/**
+ * ubrt_unregister_gsi() - Unregistering UBC's interrupt into the kernel via ACPI
+ * @hwirq: GSI IRQ number reported by BIOS
+ */
+void ubrt_unregister_gsi(u32 hwirq);
+
+/**
+ * ub_update_msi_domain() - Update the MSI domain of UBC
+ * @dev: device with ub msi domain
+ * @bus_token: DOMAIN_BUS_UB_MSI
+ *
+ * Used when booting via ACPI. The MSI domain of the UB is reported by a
+ * platform device to the driver, and this function passes the MSI domain of the
+ * platform device to the UBC.
+ *
+ * Return: 0 if success or other if failed
+ */
+int ub_update_msi_domain(struct device *dev,
+			 enum irq_domain_bus_token bus_token);
+
+/**
+ * ubrt_fwnode_set() - Associate a device's fwnode with an UBRT node
+ * @index: Index of the UBRT node within its type.
+ * @type: Type of the UBRT node.
+ * @fwnode: fwnode_handle associated with a device.
+ *
+ * Finds an UBRT node by index and type, and associates the device's
+ * fwnode with the UBRT node. This allows subsequent lookups of
+ * the UBRT node using the device's fwnode.
+ *
+ * Return: 0 if success, or error code if failed.
+ */
+int ubrt_fwnode_set(u32 index, enum ubrt_node_type type,
+		    struct fwnode_handle *fwnode);
+
+/**
+ * ubrt_fwnode_get_by_idx() - Get an UBRT node by index and type
+ * @index: Index of the UBRT node within its type.
+ * @type: Type of the UBRT node.
+ *
+ * Finds an UBRT node by index and type.
+ *
+ * Return: Pointer to the struct ubrt_fwnode if found, or NULL if not found.
+ */
+struct ubrt_fwnode *ubrt_fwnode_get_by_idx(u32 index,
+					   enum ubrt_node_type type);
+
+/**
+ * ubrt_fwnode_get() - Get an UBRT node by fwnode_handle
+ * @fwnode: fwnode_handle associated with a device.
+ *
+ * Finds an UBRT node by the device's fwnode.
+ *
+ * Return: Pointer to struct ubrt_fwnode if found, or NULL if not found.
+ */
+struct ubrt_fwnode *ubrt_fwnode_get(struct fwnode_handle *fwnode);
+
+/**
+ * ubrt_fwnode_get_count() - Get the count of UBRT nodes of a specific type
+ * @type: Type of the UBRT node.
+ *
+ * Return: The count of UBRT nodes of the specified type.
+ */
+u32 ubrt_fwnode_get_count(enum ubrt_node_type type);
+
+/**
+ * ubrt_fwnode_add() - Add an UBRT node to the ubrt_fwnode_list
+ * @node: Pointer to the UBRT node.
+ * @index: Index of the UBRT node within its type.
+ * @size: Size of the UBRT node.
+ * @type: Type of the UBRT node.
+ *
+ * Adds an UBRT node to the ubrt_fwnode_list, associating it with the
+ * specified index and type.
+ *
+ * Return: 0 if success, or error code if failed.
+ */
+int ubrt_fwnode_add(void *node, u32 index, int size, enum ubrt_node_type type);
+
+/**
+ * ubrt_fwnode_del() - Delete an UBRT node from the ubrt_fwnode_list
+ * @index: Index of the UBRT node within its type.
+ * @type: Type of the UBRT node.
+ */
+void ubrt_fwnode_del(u32 index, enum ubrt_node_type type);
+
+/**
+ * ubrt_fwnode_del_all() - Clear the ubrt_fwnode_list
+ *
+ * Clears all UBRT nodes from the ubrt_fwnode_list.
+ */
+void ubrt_fwnode_del_all(void);
+
+#if IS_ENABLED(CONFIG_UB_UBFI)
+extern struct list_head ubc_list;
+extern u32 ubc_eid_start;
+extern u32 ubc_eid_end;
+extern u32 ubc_cna_start;
+extern u32 ubc_cna_end;
+extern u8 ubc_feature;
+
+/**
+ * ubrt_iommu_get_resv_regions() - Get reserved memory regions from UBRT table.
+ * @dev: Pointer to the device.
+ * @list: List used to store reserved memory regions.
+ *
+ * In virtualization scenarios, the host OS should describe the available
+ * memory regions for UBC to the guest OS through the UB Reserved Memory
+ * information table in the UBRT table. This information table is reported by
+ * the VM BIOS; the physical machine BIOS does not need to report this
+ * information table.
+ */
+void ubrt_iommu_get_resv_regions(struct device *dev, struct list_head *list);
+
+#else /* !CONFIG_UB_UBFI */
+
+static inline
+void ubrt_iommu_get_resv_regions(struct device *dev, struct list_head *list) {}
+
+#endif /* CONFIG_UB_UBFI */
+
+#endif /* _UB_UBFI_UBFI_H_ */
-- 
2.43.0

